### ISSUE TITLE
Pipeline Resources Redesign

### ISSUE THREAD

**Issue by sbwsg** [tektoncd/pipeline#1673]
# What is this thing?

This is a tracking issue / epic for all of the ongoing pipeline resources investigation, design and proof-of-concept work.

# Existing Design Work and Feature Requests

* [Volumes, Tasks and PipelineResources overlap](https://github.com/tektoncd/pipeline/issues/1272)
* [Resources, PVCs and Chaining](https://github.com/tektoncd/pipeline/issues/1076)
* [PipelineResource Extensibility](https://github.com/tektoncd/pipeline/issues/238)
* [Tekton Creates PVCs Unnecessarily, which can be slow](https://github.com/tektoncd/pipeline/issues/937)
* [Pipeline Resource Immutability](https://github.com/tektoncd/pipeline/issues/1279)
* [Dynamic Output Resources](https://github.com/tektoncd/pipeline/issues/1180)
* [Strong Typing for Pipeline Resources](https://github.com/tektoncd/pipeline/issues/1282)
* [Optional Pipeline Resources](https://github.com/tektoncd/pipeline/issues/1710)
* [Abstract Task and Nested Tasks](https://github.com/tektoncd/pipeline/issues/1796)
* [Add Hooks into Pipeline Lifecycle](https://github.com/tektoncd/pipeline/issues/1460)
* [Allow for TaskReference Interpolation within Pipeline](https://github.com/tektoncd/pipeline/issues/1367)

# Existing PipelineResource Problems

* [Implicit Git & Image Authentication](https://github.com/tektoncd/pipeline/issues/1712)
* [Create git output that updates git](https://github.com/tektoncd/pipeline/issues/1280)
* [.ssh configs etc are not available when running manual git commands in other steps](https://github.com/tektoncd/pipeline/issues/1990)

# Design Docs Without A Home Anywhere Else

* [Pipeline Resources 2 Uber Design Doc](https://docs.google.com/document/d/1euQ_gDTe_dQcVeX4oypODGIQCAkUaMYQH5h7SaeFs44/)

**Comment by sbwsg** [comment]
PipelineResources have a big surface area so we've started to break the constituent pieces into their own deliverables. We've had good success and feedback so far as these features have been added.  [Workspaces](https://github.com/tektoncd/pipeline/blob/master/docs/taskruns.md#workspaces) and [Task Results](https://github.com/tektoncd/pipeline/blob/master/docs/tasks.md#results) are two big chunks of work that support use cases we previously covered with PipelineResources.

As we pull these pieces into their own designs and implementations we're also thinking about how we could, in the future, package them all back up into something like PipelineResources that uses all these features but limits the configuration that users have to do.  Certainly this won't happen before beta but we're still considering possibilities here.
